src/x86/freebsd/init.c

```diff
@@ -0,0 +1,402 @@
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <cpuinfo.h>
+#include <cpuinfo/internal-api.h>
+#include <cpuinfo/log.h>
+#include <freebsd/api.h>
+#include <x86/api.h>
+
+static inline uint32_t max(uint32_t a, uint32_t b) { return a > b ? a : b; }
+
+static inline uint32_t bit_mask(uint32_t bits) {
+	return (UINT32_C(1) << bits) - UINT32_C(1);
+}
+
+void cpuinfo_x86_freebsd_init(void) {
+	struct cpuinfo_processor *processors = NULL;
+	struct cpuinfo_core *cores = NULL;
+	struct cpuinfo_cluster *clusters = NULL;
+	struct cpuinfo_package *packages = NULL;
+	struct cpuinfo_cache *l1i = NULL;
+	struct cpuinfo_cache *l1d = NULL;
+	struct cpuinfo_cache *l2 = NULL;
+	struct cpuinfo_cache *l3 = NULL;
+	struct cpuinfo_cache *l4 = NULL;
+
+	struct cpuinfo_freebsd_topology freebsd_topology =
+	    cpuinfo_freebsd_detect_topology();
+	if (freebsd_topology.packages == 0) {
+		cpuinfo_log_error("failed to detect topology");
+		goto cleanup;
+	}
+	processors =
+	    calloc(freebsd_topology.threads, sizeof(struct cpuinfo_processor));
+	if (processors == NULL) {
+		cpuinfo_log_error(
+		    "failed to allocate %zu bytes for descriptions of %" PRIu32
+		    " logical processors",
+		    freebsd_topology.threads * sizeof(struct cpuinfo_processor),
+		    freebsd_topology.threads);
+		goto cleanup;
+	}
+	cores = calloc(freebsd_topology.cores, sizeof(struct cpuinfo_core));
+	if (cores == NULL) {
+		cpuinfo_log_error(
+		    "failed to allocate %zu bytes for descriptions of %" PRIu32
+		    " cores",
+		    freebsd_topology.cores * sizeof(struct cpuinfo_core),
+		    freebsd_topology.cores);
+		goto cleanup;
+	}
+	/* On x86 a cluster of cores is the biggest group of cores that shares a
+	 * cache. */
+	clusters =
+	    calloc(freebsd_topology.packages, sizeof(struct cpuinfo_cluster));
+	if (clusters == NULL) {
+		cpuinfo_log_error(
+		    "failed to allocate %zu bytes for descriptions of %" PRIu32
+		    " core clusters",
+		    freebsd_topology.packages * sizeof(struct cpuinfo_cluster),
+		    freebsd_topology.packages);
+		goto cleanup;
+	}
+	packages =
+	    calloc(freebsd_topology.packages, sizeof(struct cpuinfo_package));
```

Comment:
Is packages the reason you're wanting a branch from linux?
The latest linux kernel on android and riscv requires packages support so it would be best to use common code to avoid freebsd diverging in the future.